src/main/clojure/aerospike_clj/client.clj

```diff
@@ -172,390 +105,367 @@
-         ((:transcoder conf identity) data)
-         (policy/set-policy (get-client db) expiration)
-         set-name)))
-
-(defn create
-  "`put` with a create-only policy"
-  ([db index set-name data expiration]
-   (create db index set-name data expiration {}))
-  ([db index set-name data expiration conf]
-   (_put db
-         index
-         ((:transcoder conf identity) data)
-         (policy/create-only-policy (get-client db) expiration)
-         set-name)))
-
-(defn replace-only
-  "`put` with a replace-only policy"
-  ([db index set-name data expiration]
-   (replace-only db index set-name data expiration {}))
-  ([db index set-name data expiration conf]
-   (_put db
-         index
-         ((:transcoder conf identity) data)
-         (policy/replace-only-policy (get-client db) expiration)
-         set-name)))
-
-(defn update
-  "Writing a new value for the key `index`.
-  Generation: the expected modification count of the record (i.e. how many times was it
-  modified before my current action). Pass a function in `(:trascoder conf)` to modify
-  `data` before it is sent to the DB."
-  ([db index set-name new-record generation new-expiration]
-   (update db index set-name new-record generation new-expiration {}))
-  ([db index set-name new-record generation new-expiration conf]
-   (_put db
-         index
-         ((:transcoder conf identity) new-record)
-         (policy/update-policy (get-client db) generation new-expiration)
-         set-name)))
-
-(defn add-bins
-  "Add bins to an existing record without modifying old data. The `new-data` must be a
-  Clojure map."
-  ([db index set-name ^IPersistentMap new-data new-expiration]
-   (add-bins db index set-name new-data new-expiration {}))
-  ([db index set-name ^IPersistentMap new-data new-expiration conf]
-   (_put db
-         index
-         ((:transcoder conf identity) new-data)
-         (policy/update-only-policy (get-client db) new-expiration)
-         set-name)))
-
-(defn touch
-  "Updates the ttl of the record stored under at `index` to `expiration` seconds from now.
-  Expects records to exist."
-  [db index set-name expiration]
-  (let [client (get-client db index)
-        op-future (p/deferred)
-        start-time (System/nanoTime)]
-    (.touch ^AerospikeClient client
-            ^EventLoop (.next ^EventLoops (:el db))
+    (register-events op-future client-events :write index start-time)))
+
+(defrecord SimpleAerospikeClient [client
+                                  el
+                                  hosts
+                                  dbns
+                                  client-events
+                                  close-event-loops?]
+  pt/AerospikeReadOps
+  (get-single [this index set-name]
+    (pt/get-single this index set-name {} [:all]))
+
+  (get-single [this index set-name conf]
+    (pt/get-single this index set-name conf [:all]))
+
+  (get-single [_this index set-name conf bin-names]
+    (let [op-future  (p/deferred)
```

Comment:
You have this ritual that repeats itself in several places - create a future, get the current time, do some operation, and finally register events.
Maybe you can extract the operations into functions and remain with a generic function that just executes operations.
Of course, this would probably only work if you opt to use a single map argument instead of multiple varying arguments.

Reply:
Since these functions are internal to the library and not public to its users I don't see added value in generalizing here. On top of that, generalizing would require wrapping up parameters in a map, as you said, and unwrapping them in the implementation, which is a small bit of runtime and GC we don't gain much from spending.